liesel/band.py

```diff
@@ -0,0 +1,232 @@
+from functools import partial
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+import scipy.sparse as sps
+
+
+# @jax.jit not useful, f is compiled inside scan anyway, this makes it possible to return int.
+def bandwidth(full) -> int:
+    """
+    Returns the bandwidth p of `full` as an integer.
+
+    ## Parameters
+
+    - `full`: A full (n x n) matrix.
+    """
+
+    def f(j, row):  # Iterate over Row
+        def g(i, val):  # Iterate over value in Row
+            return i + 1, jax.lax.cond(val != 0, jnp.abs, lambda _: -1, i - j)
+
+        _, L = jax.lax.scan(g, 0, row)
+        return j + 1, L
+
+    _, b = jax.lax.scan(f, 0, full)
+    return int(jnp.max(b))
+
+
+@partial(jax.jit, static_argnames="bw")
+def to_band(full, bw, lower=True):
+    """
+    ## Parameters
+
+    - `full`: A full (n x n) matrix.
+    - `bw`: The bandwidth p.
+    - `lower`: Whether to return the (p x n) band in lower-diagonal ordered form.
+    """
+    n = full.shape[0]
+    bw += 1
+
+    # We only access the upper triangular. If we get a Matrix in lower-triangular (L) form, we convert.
+    full = jnp.where(jnp.array_equal(full, jnp.tril(full)), full.T, full)
+
+    # TODO: lax.fori_loop here?
```

Comment:
If possible, using a `fori_loop` would be great. It will probably reduce the compile time

Reply:
Tried, but hard to implement without a double for-loop. `jnp.diag(full, k)` can't be compiled at this level and I had to implement extracting the diagonal manually. My approach is commented out. Would this solution be better? Maybe you can find a better solution.

Reply:
Update: I think we have to. Compilation for big ($n>1000$) matricies takes minutes when unrolling the `for` loop.

Reply:
Alright. I've added the function `to_band_2` to the module, which uses a different approach. Could you evaluate and benchmark it, and if it works well, replace `to_band` with it?